sdk/src/main/java/com/mapbox/maps/NativeObserver.kt

```diff
@@ -0,0 +1,441 @@
+    }
+  }
+
+  fun addOnSourceRemovedListener(onSourceRemovedListener: OnSourceRemovedListener) {
+    if (onSourceRemovedListeners.isEmpty()) {
+      handler.post {
+        observable.get()?.subscribe(this, listOf(MapEvents.SOURCE_REMOVED))
+      }
+      observedEvents.add(MapEvents.SOURCE_REMOVED)
+    }
+    onSourceRemovedListeners.add(onSourceRemovedListener)
+  }
+
+  fun removeOnSourceRemovedListener(onSourceRemovedListener: OnSourceRemovedListener) {
+    onSourceRemovedListeners.remove(onSourceRemovedListener)
+    if (onSourceRemovedListeners.isEmpty()) {
+      handler.post {
+        observable.get()?.unsubscribe(this, listOf(MapEvents.SOURCE_REMOVED))
+      }
+      observedEvents.remove(MapEvents.SOURCE_REMOVED)
+    }
+  }
+
+  // Style events
+  fun addOnStyleFullyLoadedListener(onStyleFullyLoadedListener: OnStyleFullyLoadedListener) {
+    if (onStyleFullyLoadedListeners.isEmpty()) {
+      handler.post {
+        observable.get()?.subscribe(this, listOf(MapEvents.STYLE_FULLY_LOADED))
+      }
+      observedEvents.add(MapEvents.STYLE_FULLY_LOADED)
+    }
+    onStyleFullyLoadedListeners.add(onStyleFullyLoadedListener)
+  }
+
+  fun removeOnStyleFullyLoadedListener(onStyleFullyLoadedListener: OnStyleFullyLoadedListener) {
+    onStyleFullyLoadedListeners.remove(onStyleFullyLoadedListener)
+    if (onStyleFullyLoadedListeners.isEmpty()) {
+      handler.post {
+        observable.get()?.unsubscribe(this, listOf(MapEvents.STYLE_FULLY_LOADED))
+      }
+      observedEvents.remove(MapEvents.STYLE_FULLY_LOADED)
+    }
+  }
+
+  fun addOnStyleImageMissingListener(onStyleImageMissingListener: OnStyleImageMissingListener) {
+    if (onStyleImageMissingListeners.isEmpty()) {
+      handler.post {
+        observable.get()?.subscribe(this, listOf(MapEvents.STYLE_IMAGE_MISSING))
+      }
+      observedEvents.add(MapEvents.STYLE_IMAGE_MISSING)
+    }
+    onStyleImageMissingListeners.add(onStyleImageMissingListener)
+  }
+
+  fun removeOnStyleImageMissingListener(onStyleImageMissingListener: OnStyleImageMissingListener) {
+    onStyleImageMissingListeners.remove(onStyleImageMissingListener)
+    if (onStyleImageMissingListeners.isEmpty()) {
+      handler.post {
+        observable.get()?.unsubscribe(this, listOf(MapEvents.STYLE_IMAGE_MISSING))
+      }
+      observedEvents.remove(MapEvents.STYLE_IMAGE_MISSING)
+    }
+  }
+
+  fun addOnStyleImageUnusedListener(onStyleImageUnusedListener: OnStyleImageUnusedListener) {
+    if (onStyleImageUnusedListeners.isEmpty()) {
+      handler.post {
+        observable.get()?.subscribe(this, listOf(MapEvents.STYLE_IMAGE_REMOVE_UNUSED))
+      }
+      observedEvents.add(MapEvents.STYLE_IMAGE_REMOVE_UNUSED)
+    }
+    onStyleImageUnusedListeners.add(onStyleImageUnusedListener)
+  }
+
+  fun removeOnStyleImageUnusedListener(onStyleImageUnusedListener: OnStyleImageUnusedListener) {
+    onStyleImageUnusedListeners.remove(onStyleImageUnusedListener)
```

Comment:
There are so many similar codes, maybe we could extract them into one method and use `MapEvents` as a param. For example `internalRemoveListener(onStyleImageUnusedListener, MapEvents.STYLE_IMAGE_REMOVE_UNUSED)`

And for add interface  `internalAddListener(onStyleImageUnusedListener,MapEvents.STYLE_IMAGE_REMOVE_UNUSE)` . 

We could also put all listeners into a map with `MapEvents` as the key to quick access them by key. BTW, it will only need a `forEach` in `clearListeners()`

Reply:
Moved the `subscribeNewEvent` and `unsubscribeUnusedEvent` logic out, so that the redundant code is reduced, however to avoid force type casting, I prefer to leave out the hashmap approach.